### PR TITLE
Add `Html.GetSelectList()` helpers

### PR DESCRIPTION
**Pull request by dougbu**
- #438 part 2/2

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs [hunk 1]
```diff
@@ -351,6 +351,35 @@ public string DisplayText(string expression)
...
         /// <inheritdoc />
+        public IEnumerable<SelectListItem> GetSelectList<TEnum>() where TEnum : struct
+        {
+            var type = typeof(TEnum);
+            var metadata = MetadataProvider.GetMetadataForType(type);
+            if (!metadata.IsEnum || metadata.IsFlagsEnum)
+            {
+                throw new ArgumentException(
+                    Resources.FormatHtmlHelper_TypeNotSupported(type.FullName),
+                    nameof(TEnum));
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs]: This should be a better string and make it clear that this is for enums.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs [hunk 2]
```diff
@@ -1016,5 +1045,22 @@ protected virtual string GenerateValue(string expression, object value, string f
         {
             return _htmlGenerator.GetClientValidationRules(ViewContext, modelExplorer, expression);
         }
+
+        protected virtual IEnumerable<SelectListItem> GetSelectList([NotNull] ModelMetadata metadata)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs]: protected is like public, should we validate here? or perhaps name this something like `GetEnumSelectList`?

File: src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs [hunk 3]
```diff
@@ -374,7 +375,35 @@ public interface IHtmlHelper
...
+        IEnumerable<ModelClientValidationRule> GetClientValidationRules(
+            ModelExplorer modelExplorer,
+            string expression);
+
+        /// <summary>
+        /// Returns a select list for the given <typeparamref name="TEnum"/>.
+        /// </summary>
+        /// <typeparam name="TEnum">Type to generate a select list for.</typeparam>
+        /// <returns>
+        /// An IEnumerable{SelectListItem} containing the select list for the given <typeparamref name="TEnum"/>.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs]: should IEnumerable be in a `<see ... />`? (you're the expert on this).

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs [hunk 4]
```diff
@@ -858,6 +864,352 @@ public void ListBoxFor_WithUnreleatedExpression_GeneratesExpectedValue()
...
+        {
+            // Arrange
+            var metadataProvider = TestModelMetadataProvider.CreateDefaultProvider();
+            var htmlHelper = new TestHtmlHelper(metadataProvider);
+
+            // Act & Assert
+            ExceptionAssert.ThrowsArgument(
+                () => htmlHelper.GetSelectList<EnumWithFlags>(),
+                "TEnum",
+                $"The type '{ typeof(EnumWithFlags).FullName }' is not supported.");
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs]: 1538 PranavPoints have been awarded :+1:

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs [hunk 5]
```diff
@@ -858,6 +864,352 @@ public void ListBoxFor_WithUnreleatedExpression_GeneratesExpectedValue()
...
+            // Arrange
+            var metadataProvider = TestModelMetadataProvider.CreateDefaultProvider();
+            var metadata = metadataProvider.GetMetadataForType(typeof(EnumWithFields));
+            var htmlHelper = new TestHtmlHelper(metadataProvider);
+
+            // Act
+            var result = htmlHelper.GetSelectList<EnumWithFields>();
+
+            // Assert
+            Assert.Equal(metadata.ModelType, htmlHelper.Metadata.ModelType);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs]: wait, how does this happen? I'm confused about how `htmlHelper.Metadata` get its type

Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs]: Oh I see, it's  capturing thing you built for the tests. Carry on then.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs]: hmm, think I named this before decided to close #602.  without `bool?`, I can't think of another type this method could support.  I'll rename all the overloads to `GetEnumSelectList`.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs (hunk 3 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs]: good catch

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs [hunk 6]
```diff
@@ -858,6 +864,352 @@ public void ListBoxFor_WithUnreleatedExpression_GeneratesExpectedValue()
...
+            var htmlHelper = new TestHtmlHelper(metadataProvider);
+
+            // Act
+            var result = htmlHelper.GetSelectList(typeof(EnumWithFields));
+
+            // Assert
+            Assert.Equal(metadata.ModelType, htmlHelper.Metadata.ModelType);
+
+            Assert.Same(htmlHelper.SelectListItems, result);            // No replacement of the underlying List
+            VerifySelectList(htmlHelper.CopiedSelectListItems, result); // No change to the (mutable) items
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs]: I really don't understand what this is doing. This test (and `GetSelectListTEnum_WrapsGetSelectListModelMetadata`) just seems to me to verify the class you wrote to test this feature.

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs [hunk 7]
```diff
@@ -858,6 +864,352 @@ public void ListBoxFor_WithUnreleatedExpression_GeneratesExpectedValue()
...
+
+            [Display(Name = "menos uno")]
+            MinusOne = -1,
+
+#if USE_REAL_RESOURCES
+            [Display(Name = nameof(Test.Resources.DisplayAttribute_Name), ResourceType = typeof(Test.Resources))]
+#else
+            [Display(Name = nameof(TestResources.DisplayAttribute_Name), ResourceType = typeof(TestResources))]
+#endif
+            MinusTwo = -2,
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs]: Should we have a case where an enum field has a `[Display(...)]` but doesn't set the name?

**Comment by rynowak** [comment]
:watch: looks mostly good, I had a few small questions

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs]: the `<exception/>` doc comments aren't good enough for you?  :smile_cat: 

how about I add `Type must be an enum that does not have an associated FlagsAttribute.` to the resource?  I'd pass in `nameof(Enum).ToLowerInvariant()` and `nameof(FlagsAttribute)`

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs (hunk 7 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs]: :ok:

File: test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Rendering/HtmlHelperSelectTest.cs]: these two tests are attempting to verify that `GetSelectList<TEnum>()` and `GetSelectList(Type)` do nothing but verification on their own i.e. they're pass-through wrappers.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/HtmlHelper.cs]: :+1:

**Comment by rynowak** [comment]
:shipit: consider me signed off after fixing the things you agreed to fix.

**Comment by dougbu** [comment]
commit 570b1e5

**Pull request merged**